gobblin-runtime/src/main/java/org/apache/gobblin/runtime/spec_store/FSSpecStore.java

```diff
@@ -98,17 +98,39 @@ public FSSpecStore(Config sysConfig, SpecSerDe specSerDe, Optional<Logger> log)
     }
   }
 
+  public static String getFlowGroup(Path specUri){
+    return specUri.getParent().getName();
+  }
+
+  public static String getFlowName(Path specUri){
+    return Files.getNameWithoutExtension(specUri.getName());
+  }
+
   @Override
   public boolean exists(URI specUri) throws IOException {
     Preconditions.checkArgument(null != specUri, "Spec URI should not be null");
-
+    Path flowPath = new Path(specUri.getPath());
+    String flowGroup = getFlowGroup(flowPath);
+    String flowName = getFlowName(flowPath);
     FileStatus[] fileStatuses = fs.listStatus(this.fsSpecStoreDirPath);
+
     for (FileStatus fileStatus : fileStatuses) {
-      if (StringUtils.startsWith(fileStatus.getPath().getName(), specUri.toString())) {
+      if (fileStatus.isDirectory() && fileStatus.getPath().getName().equals(flowGroup)) {
+        return exists(fileStatus.getPath(), flowName);
+      } else if(!fileStatus.isDirectory() && fileStatus.getPath().getName().startsWith(flowName)) {
         return true;
       }
     }
+    return false;
+  }
 
+  private boolean exists(Path flowGroupDir, String flowName) throws IOException {
+    FileStatus[] fileStatuses = fs.listStatus(flowGroupDir);
+    for (FileStatus fileStatus : fileStatuses) {
+      if (!fileStatus.isDirectory() && fileStatus.getPath().getName().startsWith(flowName)) {
```

Comment:
I think we need to add a delimiter to the name between the name and the version. Today if you have a file abc111 we can't tell if it is version 1 of abc11 or version 11 of abc1.

Reply:
Yes, I wanted to discuss this too. But the problem is we have already written 'abc111' in the file system. We need to convert all the materialized specs to new pattern.